Introduction

This notebook will focus on analysing and preparing the data of real world car prices dataset with the aim of transforming the raw data into a clean and structured form. The dataset contains information about the different types of cars and their prices alongside a mixture of categorical attributes  describing their specifications, characteristics and market factors. 

The primary objective of this analysis is to critically assess the structure and quality of the data and apply appropriate processing techniques that includes identifying missing values or duplicated records, correcting data types and transforming variables where necessary. 


In [1]:
#Loading the Dataset

import pandas as pd        #Pandas are used for handling data structures like tables and performs data cleaning tasks
import numpy as np         #NumPY provides numerical operations thjat are required when working with real world datasets 

df= pd.read_csv("CarPrices2026.csv")  # This loads the dataset into a Pandas dataframe to enable analysis.
df.head()                             # Displaying the first few rows helps verify that the file has read correctly and the values appear in expected formats



,name,year,selling_price,km_driven,fuel,seller_type,transmission,owner,mileage(km/ltr/kg),engine,max_power,seats
0,Maruti Swift Dzire VDI,2014,450000,145500,Diesel,Individual,Manual,First Owner,23.40,1248.0,74,5.0
1,Skoda Rapid 1.5 TDI Ambition,2014,370000,120000,Diesel,Individual,Manual,Second Owner,21.14,1498.0,103.52,5.0
2,Honda City 2017-2020 EXi,2006,158000,140000,Petrol,Individual,Manual,Third Owner,17.70,1497.0,78,5.0
3,Hyundai i20 Sportz Diesel,2010,225000,127000,Diesel,Individual,Manual,First Owner,23.00,1396.0,90,5.0
4,Maruti Swift VXI BSIII,2007,130000,120000,Petrol,Individual,Manual,First Owner,16.10,1298.0,88.2,5.0


In [2]:
# Data Inspection

df.info()         # This provides an overview of the dataset by examining the column names, data types and non missing value counts. 



# Currently it shows that the dataset has 8,130 rows and 12 columns. 
# Columns that contain no missing values are Name, Year, Selling Price, Km driven, Fuel, Seller type, Transmission and Owner.
# Columns with missing values are Mileage, Engine, Max power and seats.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8130 entries, 0 to 8129
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   name                8130 non-null   object 
 1   year                8130 non-null   int64  
 2   selling_price       8130 non-null   int64  
 3   km_driven           8130 non-null   int64  
 4   fuel                8130 non-null   object 
 5   seller_type         8130 non-null   object 
 6   transmission        8130 non-null   object 
 7   owner               8130 non-null   object 
 8   mileage(km/ltr/kg)  7909 non-null   float64
 9   engine              7909 non-null   float64
 10  max_power           7915 non-null   object 
 11  seats               7909 non-null   float64
dtypes: float64(3), int64(3), object(6)
memory usage: 762.3+ KB


In [3]:
# Descriptive Analysis

df.describe()      # Using this provides an overiew of the dataset, including measuress such as the mean, standard deviation, minimum and maximum values
                   # By doing so it allows us to identify and detect anomalies, data entry errors and outliers. 



,year,selling_price,km_driven,mileage(km/ltr/kg),engine,seats
count,8130.000000,8.130000e+03,8.130000e+03,7909.000000,7909.000000,7909.000000
mean,2013.423985,6.381997e+05,6.982325e+04,19.419925,1458.597168,5.414465
std,25.167622,8.061674e+05,5.654459e+04,4.037382,503.861229,0.968353
min,200.000000,2.999900e+04,1.000000e+00,0.000000,624.000000,-5.000000
25%,2011.000000,2.549990e+05,3.500000e+04,16.780000,1197.000000,5.000000
50%,2015.000000,4.500000e+05,6.000000e+04,19.300000,1248.000000,5.000000
75%,2017.000000,6.750000e+05,9.800000e+04,22.320000,1582.000000,5.000000
max,2050.000000,1.000000e+07,2.360457e+06,42.000000,3604.000000,14.000000


In [4]:
# Missing Values Check

df.isnull().sum()   # Identifies the missing values 



name                    0
year                    0
selling_price           0
km_driven               0
fuel                    0
seller_type             0
transmission            0
owner                   0
mileage(km/ltr/kg)    221
engine                221
max_power             215
seats                 221
dtype: int64

In [5]:
# Duplicate Value Check

df.duplicated().sum()

np.int64(1204)

In [6]:
# Unique Value Analysis

# This function evaluates the variability of each feature by showing the number of distinct values as a percentage of total observations.
# Helping identify features that provide limited information due to low variability. 

unique_counts = df.nunique()
unique_percentage = (unique_counts / len(df)) *100
unique_percentage.sort_values()


transmission           0.024600
seller_type            0.036900
fuel                   0.049200
owner                  0.061501
seats                  0.135301
year                   0.393604
engine                 1.488315
max_power              3.936039
mileage(km/ltr/kg)     4.686347
selling_price          8.327183
km_driven             11.328413
name                  25.313653
dtype: float64

The results show us that some columns have very few different values but in this dataset's case it is expected. Features like transmission, fuel type, seller type and ownership fall into a small number of categories but still provide useful information about each car. Having these values does not mean that the data is poor or unhelpful but it reflects how cars are classified in the real world. Therefore there is no need to remove any of these features and the columns are kept for further analysis.

Addressing the Missing Values

During the missing values check most of the dataset is complete, but four of the columns have missing information. The mileage, engine size, max power and number of seats each have 221 missing values while all the other columns have full data. Since these features are important aspects of a car, removing the rows with missing values will affect the analysis because the information that will be removed are useful features of a car. Instead, these missing values need to be analysed carefully so that the dataset stays complete and useful for further analysis.


In [7]:
# Handling the Missing Values


df["max_power"] = df["max_power"].str.replace("bhp","",) # It is essential that the value of Max power is changed from an object to float as it will not allow calculations to be performed correctly when the data is stored as text rather than a numerical value.  
df["max_power"] = df["max_power"].str.strip() # Removes any leading or trailing spaces
df["max_power"] = df["max_power"].replace("",np.nan) # replaces empty strings with NaN
df["max_power"] = df["max_power"].astype(float) # Applying this code ensures that the value is turned from an object to a float to allow calculations.              


# Missing values were handled by using median as the calculation because it provides a central value that is less affected by outliers in the data.

df["max_power"] = df["max_power"].fillna(df["max_power"].median())
df["mileage(km/ltr/kg)"] = df["mileage(km/ltr/kg)"].fillna(df["mileage(km/ltr/kg)"].median())
df["engine"] = df["engine"].fillna(df["engine"].median())
df["seats"] = df["seats"].fillna(df["seats"].median())


df.isnull().sum() # This tells us now that there are no longer missing values and that all the columns are turned into the correct format. 
               

name                  0
year                  0
selling_price         0
km_driven             0
fuel                  0
seller_type           0
transmission          0
owner                 0
mileage(km/ltr/kg)    0
engine                0
max_power             0
seats                 0
dtype: int64

In [8]:
df.info() # This final check shows that there are no incomplete records, missing values have successfully been addressed and that the features are stored correctly in its appropriate data type. . 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8130 entries, 0 to 8129
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   name                8130 non-null   object 
 1   year                8130 non-null   int64  
 2   selling_price       8130 non-null   int64  
 3   km_driven           8130 non-null   int64  
 4   fuel                8130 non-null   object 
 5   seller_type         8130 non-null   object 
 6   transmission        8130 non-null   object 
 7   owner               8130 non-null   object 
 8   mileage(km/ltr/kg)  8130 non-null   float64
 9   engine              8130 non-null   float64
 10  max_power           8130 non-null   float64
 11  seats               8130 non-null   float64
dtypes: float64(4), int64(3), object(5)
memory usage: 762.3+ KB


In [9]:
# Defining Target and Features

y = df["selling_price"] # Target variable

x = df.drop(["selling_price","name"], axis=1) # dropping columns that are not useful

print(x.shape)
print(y.shape)


(8130, 10)
(8130,)


In [10]:
# Encoding categorical variables

x = pd.get_dummies(x, drop_first = True) # Converts text columns to numeric columns for machine learning

print(x.shape)

(8130, 16)


In [11]:
# Train test split

from sklearn.model_selection import train_test_split

# Dataset is split into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(
    x,y, test_size=0.2, random_state = 99
)

print(x_train.shape, x_test.shape) # checks the shape of the data

(6504, 16) (1626, 16)


In [12]:
# linear Regression

from sklearn.linear_model import LinearRegression

lr = LinearRegression() # Creates the model
lr.fit(x_train , y_train) # trains the model using the data

y_pred_lr = lr.predict(x_test) # predicts on test data

In [13]:
# Evaulating Lineare Regression

from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred_lr) # calculates the mean squared error
rmse = np.sqrt(mse) # converts MSE to RMSE by square root

print("Linear Regression results:")
print("MAE:", mean_absolute_error (y_test, y_pred_lr)) # Average error
print("RMSE:", rmse) # Penalises large errors
print("R2:", r2_score(y_test, y_pred_lr)) # model performance

Linear Regression results:
MAE: 336410.43855757936
RMSE: 1851269.1429030474
R2: -4.34979310108311


In [14]:
# Random Forest Regressor

from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(random_state = 99) # Creates the model
rf.fit(x_train, y_train) # Trains the model

y_pred_rf = rf.predict(x_test) # Predicts the test data

In [15]:
# Evaluates the random forest

mse_rf = mean_squared_error(y_test, y_pred_rf)
rmse_rf = np.sqrt (mse_rf)

print("Random Forest results:")
print("MAEl", mean_squared_error(y_test, y_pred_rf))
print("RMSE:", rmse_rf)
print("R2", r2_score(y_test, y_pred_rf))

Random Forest results:
MAEl 18974600417.0721
RMSE: 137748.32273778183
R2 0.970380992574823


In [16]:
# Gradient boosting

from sklearn.ensemble import GradientBoostingRegressor

gb = GradientBoostingRegressor(random_state = 99) # Creates the Gradient boosting model
gb.fit(x_train, y_train) # Trains the model

y_pred_gb = gb.predict(x_test) # Predicts the test data

In [17]:
# Evaluates the Gradient Boosting

mse_gb = mean_squared_error(y_test, y_pred_gb)
rmse_gb = np.sqrt(mse_gb)

print("Gradient Boosting Results:")
print("MAE:", mean_absolute_error(y_test, y_pred_gb))
print("RMSE:", rmse_gb)
print("R2:", r2_score(y_test, y_pred_gb))

Gradient Boosting Results:
MAE: 92138.42912825797
RMSE: 160308.30382582542
R2: 0.9598847133675945


In [18]:
# Comparison Table


# Creates comparison table of all models
results = pd.DataFrame({
    "Model": ["Linear Regression", "Random Forest", "Gradient Boosting"],
    "R2": [
        r2_score(y_test, y_pred_lr),
        r2_score(y_test, y_pred_rf),
        r2_score(y_test, y_pred_gb)
    ],
    "RMSE": [
        rmse,
        rmse_rf,
        rmse_gb
    ]
})

print(results)

               Model        R2          RMSE
0  Linear Regression -4.349793  1.851269e+06
1      Random Forest  0.970381  1.377483e+05
2  Gradient Boosting  0.959885  1.603083e+05
